File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.java
class Solution {
    public int longestPalindrome(String[] words) {
        int [][] count = new int[26][26];
        int te = 0;
        for(String w : words){
            int a = w.charAt(0) - 'a';
            int b = w.charAt(1) - 'a';
            if(count[b][a] > 0){
                te += 4;
                count[b][a]--;
            }
            else
                count[a][b]++;
        }
        for(int i=0; i<26; i++){
            if(count[i][i] >0){
                te += 2;
                break;
            }
        }
        return te;
    }
}

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        map<string,int>m;
        int ans=0;
        for(int i=0;i<words.size();i++){
            string str=words[i];
            reverse(str.begin(),str.end());
            if(m.find(str)!=m.end())
            {    ans+=4;
                m[str]--;
             if(m[str]==0) m.erase(str);
            }
            else
                m[words[i]]++;
        }
        for(auto it:m){
            if(it.first[0]==it.first[1])
            {    ans+=2; break;}
            
        }
      
        return ans;
    } 
};